### PR TITLE
Add by_threshold ranking metrics regression test

### LINKED ISSUE #2154: [BUG] Ranking Evaluation Metrics Exceed 1 with "by_threshold" Relevancy Method
### Description
Hello!

I encountered an issue while evaluating the BPR (Bayesian Personalized Ranking) model with basically the same code provided in the example on a different dataset. Specifically, when using the "by_threshold" relevancy method with ranking metrics, the computed values for precision@k, ndcg@k, and map@k exceed 1, which seems incorrect. This issue does not occur when switching the relevancy method to "top_k."

### How do we replicate the issue?
I use the following parameter for BPR (all using the default seed):
```
bpr = cornac.models.BPR(
    k=200,
    max_iter=100,
    learning_rate=0.01,
    lambda_reg=0.001,
    verbose=True 
)
```
Using these evaluation
```
TOP_K = 10
threshold =50
eval_map = map_at_k(test, all_predictions, col_prediction="prediction",
                    relevancy_method='by_threshold', threshold=threshold, k=TOP_K)
eval_ndcg = ndcg_at_k(test, all_predictions, col_prediction="prediction",
                      relevancy_method='by_threshold', threshold=threshold, k=TOP_K)
eval_precision = precision_at_k(
    test, all_predictions, col_prediction="prediction",
    relevancy_method='by_threshold', threshold=threshold, k=TOP_K)
```
Here is the dataset I test on:  https://github.com/mnhqut/rec_sys-dataset/blob/main/data.csv

My result: 
MAP:		1.417529
NDCG:		1.359902
Precision@K:	2.256466
### Willingness to contribute

- [ ] Yes, I can contribute for this issue independently.
- [x ] Yes, I can contribute for this issue with guidance from Recommenders community.
- [ ] No, I cannot contribute at this time.

### PR DESCRIPTION
**Pull request by benben951**
Fixes #2154.

This adds a regression test for Python ranking metrics with `relevancy_method="by_threshold"`. The staging branch already filters predictions by threshold before applying the top-k cutoff; this test locks in that behavior so `precision_at_k`, `ndcg_at_k`, and `map_at_k` stay bounded when more than `k` items pass the threshold.

Tested:
- `python -m pytest tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_ranking_metrics_by_threshold_are_bounded tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_precision_at_k tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_ndcg_at_k tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_map_at_k -q`
- Result: `4 passed`

### CODE REVIEW

**Comment by miguelgfierro** [comment]
Hi @benben951! Your PR has a merge conflict with `staging` — both branches inserted a new test function immediately after `test_python_map_at_k` at the same line. Staging added `test_python_map_vs_map_at_k`; your PR added `test_python_ranking_metrics_by_threshold_are_bounded`.

Here's how to resolve it:

```bash
git fetch upstream staging
git rebase upstream/staging
```

When the conflict appears in `tests/unit/recommenders/evaluation/test_python_evaluation.py`, edit the file to keep **both** new functions. The resolved section should look like this:

```python
def test_python_map_vs_map_at_k(rating_true, rating_pred):
    """``map_at_k`` and ``map`` differ only in the per-user normalizer.
    ...
    """
    assert map_at_k(rating_true, rating_pred, k=10) == pytest.approx(
        map(rating_true, rating_pred, k=10), TOL
    )
    assert map_at_k(rating_true, rating_pred, k=5) > map(rating_true, rating_pred, k=5)


def test_python_ranking_metrics_by_threshold_are_bounded():
    rating_true = pd.DataFrame(
        {
            DEFAULT_USER_COL: [1, 1, 1, 1],
            DEFAULT_ITEM_COL: [1, 2, 3, 4],
            DEFAULT_RATING_COL: [1, 1, 1, 1],
        }
    )
    rating_pred = pd.DataFrame(
        {
            DEFAULT_USER_COL: [1, 1, 1, 1],
            DEFAULT_ITEM_COL: [1, 2, 3, 4],
            DEFAULT_PREDICTION_COL: [100, 99, 98, 97],
        }
    )

    assert precision_at_k(
        rating_true,
        rating_pred,
        relevancy_method="by_threshold",
        threshold=50,
        k=2,
    ) == 1.0
    assert ndcg_at_k(
        rating_true,
        rating_pred,
        relevancy_method="by_threshold",
        threshold=50,
        k=2,
    ) == 1.0
    assert map_at_k(
        rating_true,
        rating_pred,
        relevancy_method="by_threshold",
        threshold=50,
        k=2,
    ) == 1.0


def test_python_precision_at_k(rating_true, rating_pred, rating_nohit):
```

Then:

```bash
git add tests/unit/recommenders/evaluation/test_python_evaluation.py
git rebase --continue
git push --force-with-lease
```

I verified locally that `test_python_ranking_metrics_by_threshold_are_bounded` passes after this resolution.

**Comment by benben951** [comment]
Thanks @miguelgfierro! I resolved the conflict by rebasing the PR branch onto the latest `staging` state and keeping both tests in `tests/unit/recommenders/evaluation/test_python_evaluation.py`:

- `test_python_map_vs_map_at_k`
- `test_python_ranking_metrics_by_threshold_are_bounded`

I also verified the resolved section locally with:

```bash
python -m pytest tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_map_vs_map_at_k tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_ranking_metrics_by_threshold_are_bounded -q
```

Result: `2 passed`.

**Comment by benben951** [comment]
A quick note on the latest `unit-tests / get-test-groups` failure: this appears to be caused by missing repository secrets for this fork PR rather than by the test change itself.

The check annotation says the reusable workflow is missing:

```text
COMPSHARE_PRIVATE_KEY
COMPSHARE_PUBLIC_KEY
DOCKER_MIRROR_URL
PIP_INDEX_URL
```

So the job fails during workflow setup before any test steps run. The PR is currently mergeable, DCO is passing, and I verified the relevant tests locally:

```bash
python -m pytest tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_map_vs_map_at_k tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_ranking_metrics_by_threshold_are_bounded -q
```

Result: `2 passed`.

**Comment by anargyri** [comment]
> A quick note on the latest `unit-tests / get-test-groups` failure: this appears to be caused by missing repository secrets for this fork PR rather than by the test change itself.
> 
> The check annotation says the reusable workflow is missing:
> 
> ```
> COMPSHARE_PRIVATE_KEY
> COMPSHARE_PUBLIC_KEY
> DOCKER_MIRROR_URL
> PIP_INDEX_URL
> ```
> 
> So the job fails during workflow setup before any test steps run. The PR is currently mergeable, DCO is passing, and I verified the relevant tests locally:
> 
> ```shell
> python -m pytest tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_map_vs_map_at_k tests/unit/recommenders/evaluation/test_python_evaluation.py::test_python_ranking_metrics_by_threshold_are_bounded -q
> ```
> 
> Result: `2 passed`.

Can you merge staging again? I think PR #2328 has fixed this.

**Comment by benben951** [comment]
Thanks @anargyri! I merged the latest `staging` into this PR branch.

The new head is `a57c9112`, which includes #2328 (`Explicitly specify secrets to use`). The PR is still mergeable and DCO is passing.

**Comment by benben951** [comment]
﻿The rerun got past `run-tests / get-test-groups`, but the GPU matrix jobs still fail before tests execute.

From the failed `run-tests / 3.10, group_gpu` / `3.11, group_gpu` logs:

```text
Secret source: None
COMPSHARE_PRIVATE_KEY:
COMPSHARE_PUBLIC_KEY:
DOCKER_MIRROR_URL:
ERROR: {"RetCode":230,"Message":"Fail to parse parameters"}
```

So this still looks like the Compshare VM setup is receiving empty secrets for this fork PR, rather than a failure in the ranking metrics regression test itself.

**Comment by anargyri** [comment]
Could you merge staging again? There has been another fix.

**Comment by SimonYansenZhao** [comment]
Could merge the staging branch again?  We included a fix for the VM setup error.

**Comment by benben951** [comment]
﻿Thanks @anargyri and @SimonYansenZhao. I merged the latest `staging` again and pushed the updated PR branch.

New head: `fe675f3b`, which includes #2330 and the VM setup fixes. The PR is still mergeable and DCO is passing.

**Review by miguelgfierro — APPROVED** [review]
Thanks!
